File: client/src/main/java/net/hoz/netapi/client/data/DataHolderImpl.java
package net.hoz.netapi.client.data;

import com.iamceph.resulter.core.DataResultable;
import com.iamceph.resulter.core.Resultable;
import lombok.ToString;
import lombok.extern.slf4j.Slf4j;
import org.spongepowered.configurate.CommentedConfigurationNode;
import org.spongepowered.configurate.ConfigurateException;
import org.spongepowered.configurate.ConfigurationNode;
import org.spongepowered.configurate.gson.GsonConfigurationLoader;

@Slf4j
@ToString
final class DataHolderImpl implements DataHolder {
    private ConfigurationNode root;

    DataHolderImpl(String json) throws ConfigurateException {
        this.root = build(json);
    }

    @Override
    public ConfigurationNode root() {
        return root;
    }

    @Override
    public ConfigurationNode node(String key) {
        if (key == null) {
            log.trace("Something tried to get null node!");
            return CommentedConfigurationNode.root();
        }

        return root.node((Object[]) key.split("\\."));
    }

    @Override
    public Resultable update(String json) {
        try {
            this.root = build(json);
            return Resultable.ok();
        } catch (ConfigurateException e) {
            return Resultable.fail(e);
        }
    }

    @Override
    public DataResultable<String> get() {
        try {
            return DataResultable.failIfNull(GsonConfigurationLoader.builder().buildAndSaveString(root));
        } catch (Exception e) {
            return DataResultable.fail(e);
        }
    }

    private ConfigurationNode build(String json) throws ConfigurateException {
        return GsonConfigurationLoader
                .builder()
                .buildAndLoadString(json);
    }
}

File: client/src/main/kotlin/net/hoz/netapi/client/data/DataHolderImpl.kt
/*
 * Copyright 2022 hoz-network.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.hoz.netapi.client.data

import com.iamceph.resulter.core.DataResultable
import com.iamceph.resulter.core.Resultable
import com.iamceph.resulter.kotlin.dataResultable
import com.iamceph.resulter.kotlin.resultable
import org.spongepowered.configurate.ConfigurationNode
import org.spongepowered.configurate.gson.GsonConfigurationLoader

internal data class DataHolderImpl(override var root: ConfigurationNode) : DataHolder {
    constructor(json: String) : this(GsonConfigurationLoader.builder().buildAndLoadString(json))

    override fun node(key: String): ConfigurationNode = root.node(key.split("\\."))

    override fun update(input: String): Resultable = resultable { root = GsonConfigurationLoader.builder().buildAndLoadString(input) }

    override fun json(): DataResultable<String> = dataResultable { GsonConfigurationLoader.builder().buildAndSaveString(root) }
}